lib/spack/spack/database.py

```diff
@@ -0,0 +1,290 @@
+            raise SpackYAMLError("error parsing YAML database:", str(e))
+
+        if file==None:
+            return
+
+        data = {}
+        for index, sp in file['database'].items():
+            spec = Spec.from_node_dict(sp['spec'])
+            deps = sp['dependency_indices']
+            path = sp['path']
+            dep_hash = sp['hash']
+            db_entry = {'deps':deps, 'spec': spec, 'path': path, 'hash':dep_hash}
+            data[index] = db_entry
+
+        for sph in data.values():
+            for idx in sph['deps']:
+                sph['spec'].dependencies[data[idx]['spec'].name] = data[idx]['spec']
+
+        self.data = data.values()
+
+
+    def read_database(self):
+        """
+        Re-read Database from the data in the set location
+        If the cache is fresh, return immediately.
+        Implemented with mkdir locking for the database file.
+        """
+        if not self.is_dirty():
+            return
+
+        lock=0
+        while lock==0:
+            try:
+                os.mkdir(self.lock_path)
+                lock=1
+            except OSError as err:
+                pass
+
+        #The try statement ensures that a failure won't leave the
+        #database locked to other processes.
+        try:
+            if os.path.isfile(self.file_path):
+                with open(self.file_path,'r') as f:
+                    self.from_yaml(f)
+            else:
+            #The file doesn't exist, construct empty data.
+                self.data = []
+        except:
+            os.rmdir(self.lock_path)
+            raise
+
+        os.rmdir(self.lock_path)
+
+
+    def write_database_to_yaml(self,stream):
+        """
+        Replace each spec with its dict-node form
+        Then stream all data to YAML
+        """
+        node_list = []
+        spec_list = [sph['spec'] for sph in self.data]
+
+        for sph in self.data:
+            node = {}
+            deps = []
+            for name,spec in sph['spec'].dependencies.items():
+                deps.append(spec_list.index(spec))
+            node['spec']=Spec.to_node_dict(sph['spec'])
+            node['hash']=sph['hash']
+            node['path']=sph['path']
+            node['dependency_indices']=deps
+            node_list.append(node)
+
+        node_dict = dict(enumerate(node_list))
+        return yaml.dump({ 'database' : node_dict},
+                         stream=stream, default_flow_style=False)
+
+
+    def write(self):
+        """
+        Write the database to the standard location
+        Implements mkdir locking for the database file
+        """
+        lock=0
+        while lock==0:
+            try:
+                os.mkdir(self.lock_path)
+                lock=1
+            except OSError as err:
+                pass
+
+        #The try statement ensures that a failure won't leave the
+        #database locked to other processes.
+        try:
+            with open(self.file_path,'w') as f:
```

Comment:
One issue with writing directly to the DB file is that this is not an atomic operation, so if you hit a disk quota or an error, you will end up with a corrupted DB.

The way you should avoid this is to make a temporary file alongside the existing file, then `move` it into place:

http://stackoverflow.com/questions/8745387/in-python-make-a-tempfile-in-the-same-directory-as-another-file

For most filesystems, `mv` does an atomic update, so you will only blow away the old DB if the new DB was successfully written.